corporate/lib/stripe.py

```diff
@@ -315,75 +317,181 @@ def stripe_get_customer(stripe_customer_id: str) -> stripe.Customer:
+        )
+        self.write_to_audit_log(AbstractRealmAuditLog.STRIPE_CARD_CHANGED, timezone_now())
+        if pay_invoices:
+            for stripe_invoice in stripe.Invoice.list(
+                collection_method="charge_automatically",
+                customer=stripe_customer_id,
+                status="open",
+            ):
+                # The stripe customer with the associated ID will get either a receipt
+                # or a "failed payment" email, but the in-app messaging could be clearer
+                # here (e.g. it could explicitly tell the user that there were payment(s)
+                # and that they succeeded or failed). Worth fixing if we notice that a
+                # lot of cards end up failing at this step.
+                stripe.Invoice.pay(stripe_invoice)
+
+    # Returns Customer instead of stripe_customer so that we don't make a Stripe
+    # API call if there's nothing to update
+    @catch_stripe_errors
+    def update_or_create_stripe_customer(self, payment_method: Optional[str] = None) -> Customer:
+        customer = self.get_customer()
+        if customer is None or customer.stripe_customer_id is None:
+            # We could do a better job of handling race conditions here, but if two
+            # people try to upgrade at exactly the same time, the main bad thing that
+            # will happen is that we will create an extra stripe customer that we can
+            # delete or ignore.
+            return self.create_stripe_customer(payment_method=payment_method)
         if payment_method is not None:
-            RealmAuditLog.objects.create(
-                realm=user.realm,
-                acting_user=user,
-                event_type=RealmAuditLog.STRIPE_CARD_CHANGED,
-                event_time=event_time,
+            self.replace_payment_method(customer.stripe_customer_id, payment_method, True)
+        return customer
+
+    def attach_discount_to_customer(self, discount: Decimal) -> None:
+        customer = self.get_customer()
+        old_discount: Optional[Decimal] = None
+        if customer is not None:
+            old_discount = customer.default_discount
+            customer.default_discount = discount
+            customer.save(update_fields=["default_discount"])
+        else:
+            customer = self.update_or_create_customer(defaults={"default_discount": discount})
+        plan = get_current_plan_by_customer(customer)
+        if plan is not None:
+            plan.price_per_license = get_price_per_license(
+                plan.tier, plan.billing_schedule, discount
             )
-        customer, created = Customer.objects.update_or_create(
-            realm=realm, defaults={"stripe_customer_id": stripe_customer.id}
+            plan.discount = discount
+            plan.save(update_fields=["price_per_license", "discount"])
+        self.write_to_audit_log(
+            event_type=AbstractRealmAuditLog.REALM_DISCOUNT_CHANGED,
```

Comment:
This would be different for a Customer with remote_server (`AbstractRealmAuditLog.REMOTE_SERVER_DISCOUNT_CHANGED` theoretically), so thinking about how to move this to the child class.

Reply:
A possible solution is to just set `AbstractRealmAuditLog.REMOTE_SERVER_DISCOUNT_CHANGED = AbstractRealmAuditLog.REALM_DISCOUNT_CHANGED` when picking values.

But probably we can do this pretty cleanly with some tiny abstraction like `self.AUDIT_LOG_CLASS.DISCOUNT_CHANGED_VALUE`.

(In any case, not a blocker for moving a bunch more stuff into this billing session abstraction).

Reply:
Added a commit (4th in the series) to create an abstraction for these audit log event types.